src/main/java/com/cdancy/bitbucket/rest/features/InsightsApi.java

```diff
@@ -0,0 +1,167 @@
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+
+@Produces(MediaType.APPLICATION_JSON)
+@RequestFilters(BitbucketAuthenticationFilter.class)
+@Path("/rest/insights/{jclouds.api-version}")
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+public interface InsightsApi {
+
+    @Named("insights:get-annotations")
+    @Documentation( {"https://docs.atlassian.com/bitbucket-server/rest/6.4.0/bitbucket-code-insights-rest.html#idp2"})
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Path("/projects/{project}/repos/{repo}/commits/{commitId}/annotations")
+    @Fallback(BitbucketFallbacks.AnnotationsResponseOnError.class)
+    @GET
+    AnnotationsResponse listAnnotations(@PathParam("project") String project,
+                                        @PathParam("repo") String repo,
+                                        @PathParam("commitId") String commitId,
+                                        @Nullable @QueryParam("externalId") String externalId,
+                                        @Nullable @QueryParam("path") String path,
+                                        @Nullable @QueryParam("severity") String severity,
+                                        @Nullable @QueryParam("type") String type);
+
+    @Named("insights:get-reports")
+    @Documentation( {"https://docs.atlassian.com/bitbucket-server/rest/6.4.0/bitbucket-code-insights-rest.html#idp4"})
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Path("/projects/{project}/repos/{repo}/commits/{commitId}/reports")
+    @Fallback(BitbucketFallbacks.InsightReportPageOnError.class)
+    @GET
+    InsightReportPage listReports(@PathParam("project") String project,
+                                  @PathParam("repo") String repo,
+                                  @PathParam("commitId") String commitId,
+                                  @Nullable @QueryParam("limit") int limit,
+                                  @Nullable @QueryParam("start") int start);
+
+    @Named("insights:get-report")
+    @Documentation( {"https://docs.atlassian.com/bitbucket-server/rest/6.4.0/bitbucket-code-insights-rest.html#idp7"})
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Path("/projects/{project}/repos/{repo}/commits/{commitId}/reports/{key}")
+    @Fallback(BitbucketFallbacks.InsightReportOnError.class)
+    @GET
+    InsightReport getReport(@PathParam("project") String project,
+                            @PathParam("repo") String repo,
+                            @PathParam("commitId") String commitId,
+                            @PathParam("key") String key);
+
+    @Named("insights:create-report")
+    @Documentation( {"https://docs.atlassian.com/bitbucket-server/rest/6.4.0/bitbucket-code-insights-rest.html#idp9"})
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Path("/projects/{project}/repos/{repo}/commits/{commitId}/reports/{key}")
+    @Fallback(BitbucketFallbacks.InsightReportOnError.class)
+    @PUT
+    InsightReport createReport(@PathParam("project") String project,
```

Comment:
:warning: Rather than using a lot of String arguments, consider using a container object for those values. (#QR10:99#)-PMD_UseObjectForClearerAPI